storage/src/journal/fixed.rs

```diff
@@ -0,0 +1,656 @@
+            let mut it = blobs.iter();
+            let mut previous_blob = it.next().unwrap().1;
+            oldest_blob = previous_blob;
+            for (_, section) in it {
+                if *section != previous_blob + 1 {
+                    return Err(Error::MissingBlob(previous_blob + 1));
+                }
+                previous_blob = *section;
+            }
+        } else {
+            debug!("no blobs found, bootstrapping with 1 empty blob");
+            let blob = runtime.open(&cfg.partition, &0u64.to_be_bytes()).await?;
+            blobs.push((blob, 0));
+        }
+
+        // Initialize metrics
+        let tracked = Gauge::default();
+        let synced = Counter::default();
+        let pruned = Counter::default();
+        {
+            let mut registry = cfg.registry.lock().unwrap();
+            registry.register("tracked", "Number of blobs", tracked.clone());
+            registry.register("synced", "Number of syncs", synced.clone());
+            registry.register("pruned", "Number of blobs pruned", pruned.clone());
+        }
+        tracked.set(blobs.len() as i64);
+
+        let blobs: Vec<B> = blobs.into_iter().map(|(blob, _)| blob).collect();
+        // truncate the last blob if it's not the expected length, which might happen from unclean shutdown.
+        let last_blob = blobs.last().unwrap();
+        let blob_len: u64 = last_blob.len().await?;
+        if blob_len % (size_of::<u32>() + N) as u64 != 0 {
+            warn!(
+                "last blob len ({}) is not a multiple of item size, truncating",
+                blob_len
+            );
+            last_blob
+                .truncate(blob_len - blob_len % (size_of::<u32>() + N) as u64)
+                .await?;
+            last_blob.sync().await?;
+        }
+
+        Ok(Self {
+            runtime,
+            cfg,
+            oldest_unpruned_blob: oldest_blob as usize,
+            blobs,
+            tracked,
+            synced,
+            pruned,
+        })
+    }
+
+    /// Sync any pending updates to disk.
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        self.synced.inc();
+        debug!("syncing journal {:?}", self.synced);
+        self.blobs
+            .last_mut()
+            .unwrap()
+            .sync()
+            .await
+            .map_err(Error::Runtime)
+    }
+
+    /// Append a new item to the journal. Return the item's position in the journal, or error if the
+    /// operation fails.
+    pub async fn append(&mut self, item: [u8; N]) -> Result<u64, Error> {
+        let mut last_blob = self.blobs.last_mut().unwrap();
+        let mut blob_len = last_blob.len().await?;
+        assert_eq!(blob_len % (size_of::<u32>() + N) as u64, 0);
+        let items_in_blob = blob_len / (size_of::<u32>() + N) as u64;
+
+        // if the blob is full we need to create the next one and use that instead
+        if items_in_blob >= self.cfg.items_per_blob {
```

Comment:
I think you should emit a debug log when we move into a new blob

Reply:
done